lib/ZnapZend.pm

```diff
@@ -535,14 +556,32 @@ my $sendRecvCleanup = sub {
         # See if there is anything remaining to clean up in child
         # datasets (if recursive snapshots are enabled) or just
         # this one dataset.
+        # Note: we apply reverse sorting by dataset names, so any
+        # "custom" child dataset snapshots (made with --inherited
+        # mode starting from mid-tree) are deleted first, parents
+        # last, and we do not lose track of mostRecentCommonSnapshot
+        # if we have to iterate up to root to find dst_X_synced.
         SRC_SET:
-        for my $srcDataSet (@$srcSubDataSets){
+        for my $srcDataSet (sort {$b cmp $a} @$srcSubDataSets){
             next if ($backupSet->{recursive} eq 'on' && $srcDataSet eq $backupSet->{src});
 
+            $self->zLog->debug('checking to clean up snapshots of source '. $srcDataSet);
+
             @snapshots = @{$self->zZfs->listSnapshots($srcDataSet, $backupSet->{snapCleanFilter})};
             $toDestroy = $self->zTime->getSnapshotsToDestroy(\@snapshots,
                          $backupSet->{srcPlanHash}, $backupSet->{tsformat}, $timeStamp);
 
+            for (@{$toDestroy}) {
+                my $snapname = $_;
```

Comment:
is this to make sure that the content of the $toDestroy array does not get changed ?

Reply:
maybe having a function that takes care of extracting the snapshot names, and returning an array would make sense here ?

Reply:
good point, this bit of code happens a few times too many :)

Reply:
Although not sure on second thought: I tracked twice the snapshots we considered (in filtered `@snapshots` and selection `@toDestroy`) with different values... I guess there was an idea to differentiate the two though not sure now, probably it did not get into code and popped out of memory at the moment. So we'd need then to merge the two resulting arrays also, instead of now just writing into one. Or pass several to one routine...

Reply:
What I mean is that instead of walking the array and running the regexp to extract the snapshot names you could create a function which does that

```perl 
sub extractSnapshotNames($array){
      my @ret;
      for (@$array){
           /\@(.+)/ and push @ret, $1;
     }
     return \@ret;
}
```

Reply:
Thanks, I'll see if a usecase remains for tracking that as a hash with various levels, or an array is okay.

Reply:
I guess the hash usage originated with another sanity check, to make sure values are unique just in case. And since we had a hash, I decided to keep it, e.g. to have speedy lookups into the keys (is this snapname in the list?)...

Reply:
Done, pushed, tested